test/functional/interface_bitcoin_cli.py

```diff
@@ -107,6 +108,23 @@ def run_test(self):
         self.log.info("Test connecting to a non-existing server")
         assert_raises_process_error(1, "Could not connect to the server", self.nodes[0].cli('-rpcport=1').echo)
 
+        self.log.info("Test rpcconnect handling of invalid ports")
+        assert_raises_process_error(1, "Invalid port provided in -rpcconnect: 127.0.0.1:notaport", self.nodes[0].cli("-rpcconnect=127.0.0.1:notaport").echo)
+        assert_raises_process_error(1, "Invalid port provided in -rpcconnect: 127.0.0.1:-1", self.nodes[0].cli("-rpcconnect=127.0.0.1:-1").echo)
+        assert_raises_process_error(1, "Invalid port provided in -rpcconnect: 127.0.0.1:0", self.nodes[0].cli("-rpcconnect=127.0.0.1:0").echo)
+
+        self.log.info("Test rpcport handling of invalid ports")
+        assert_raises_process_error(1, "Invalid port provided in -rpcport: notaport", self.nodes[0].cli("-rpcport=notaport").echo)
+        assert_raises_process_error(1, "Invalid port provided in -rpcport: -1", self.nodes[0].cli("-rpcport=-1").echo)
+        assert_raises_process_error(1, "Invalid port provided in -rpcport: 0", self.nodes[0].cli("-rpcport=0").echo)
+
+        self.log.info("Test preference of port in rpcport over rpcconnect")
+        node_rpc_port = rpc_port(self.nodes[0].index)
+        assert_raises_process_error(1, "Could not connect to the server", self.nodes[0].cli(f"-rpcconnect=127.0.0.1:{node_rpc_port}", "-rpcport=1").echo)
+        assert_equal(BLOCKS, self.nodes[0].cli("-rpcconnect=127.0.0.1:18999", f'-rpcport={node_rpc_port}').getblockcount())
+        assert_equal(BLOCKS, self.nodes[0].cli(f"-rpcconnect=127.0.0.1:{node_rpc_port}").getblockcount())
+        assert_equal(BLOCKS, self.nodes[0].cli(f'-rpcport={node_rpc_port}').getblockcount())
+
```

Comment:
Also nice to have would be a test when no port is provided that we fall back to `BaseParams().RPCPort()`, it isn't obvious to me how that could be done here though.

Reply:
Added test for default port in regtest.

Thank you for drawing attention to this!  Looks like explicitly defined `-rpcport` during the creation of the node's bitcoin.conf was causing bitcoin-cli to use the conf-provided rpcport when using rpcconnect without rpcport in command line arguments.  One of the tests (involving `f"-rpcconnect=127.0.0.1:{node_rpc_port}"`) would always pass because of this (i.e. the test wouldn't fail even if a valid but unused port is specified in rpcconnect).  This was verified by creating a temporary test `f"-rpcconnect=127.0.0.1:1234"` that should have failed but didn't.

This was remedied by using `replace_in_config()`.